providers/profitbricks/src/main/java/org/jclouds/profitbricks/compute/extensions/ProfitBricksImageExtension.java

```diff
@@ -0,0 +1,139 @@
+import org.jclouds.profitbricks.domain.Storage;
+
+import com.google.common.base.Function;
+import com.google.common.base.Predicate;
+import com.google.common.collect.Iterables;
+import com.google.common.util.concurrent.ListenableFuture;
+import com.google.common.util.concurrent.ListeningExecutorService;
+import com.google.common.util.concurrent.UncheckedTimeoutException;
+import com.google.inject.Inject;
+
+@Singleton
+public class ProfitBricksImageExtension implements ImageExtension {
+
+   @Resource
+   @Named(ComputeServiceConstants.COMPUTE_LOGGER)
+   protected Logger logger = Logger.NULL;
+
+   private final Predicate<Storage> matchBootDevice = new Predicate<Storage>() {
+      @Override
+      public boolean apply(Storage input) {
+         return input.bootDevice() == null ? false : input.bootDevice();
+      }
+   };
+
+   private final ProfitBricksApi api;
+   private final Predicate<String> snapshotAvailablePredicate;
+   private final Function<Provisionable, Image> imageTransformer;
+   private final ListeningExecutorService userExecutor;
+
+   @Inject
+   ProfitBricksImageExtension(ProfitBricksApi api,
+           @Named(TIMEOUT_SNAPSHOT_AVAILABLE) Predicate<String> snapshotAvailablePredicate,
+           Function<Provisionable, Image> imageTransformer,
+           @Named(Constants.PROPERTY_USER_THREADS) ListeningExecutorService userExecutor) {
+      this.api = api;
+      this.snapshotAvailablePredicate = snapshotAvailablePredicate;
+      this.imageTransformer = imageTransformer;
+      this.userExecutor = userExecutor;
+   }
+
+   @Override
+   public ImageTemplate buildImageTemplateFromNode(String name, String id) {
+      Server server = api.serverApi().getServer(id);
+
+      if (server == null)
+         throw new NoSuchElementException("Cannot find server with id: " + id);
+
+      List<Storage> storages = server.storages();
+      if (storages.isEmpty() || !Iterables.tryFind(storages, matchBootDevice).isPresent())
+         throw new NoSuchElementException("Server " + id + " does not contain a boot device");
+
+      return new ImageTemplateBuilder.CloneImageTemplateBuilder().nodeId(id).name(name).build();
+   }
+
+   @Override
+   public ListenableFuture<Image> createImage(ImageTemplate template) {
+      checkState(template instanceof CloneImageTemplate, "profitbricks only supports creating images through cloning.");
+      final CloneImageTemplate cloneTemplate = (CloneImageTemplate) template;
+      String serverId = cloneTemplate.getSourceNodeId();
+
+      Server server = api.serverApi().getServer(serverId);
+      Storage bootDevice = Iterables.find(server.storages(), matchBootDevice);
+
+      final Snapshot requested = api.snapshotApi().createSnapshot(
+              Snapshot.Request.creatingBuilder()
+              .storageId(bootDevice.id())
+              .name(template.getName())
+              .description(template.getName() + " (created with jclouds)")
+              .build());
+
```

Comment:
Since this is something that takes time, add a log message here saying that the image is being created.